mirascope/mcp/server.py

```diff
@@ -0,0 +1,326 @@
+
+    Returns:
+        A list of types.PromptMessage instances.
+    """
+
+    # Validate role
+    role = base_message_param.role
+    if role not in ["user", "assistant"]:
+        raise ValueError(f"invalid role: {role}")
+
+    contents = []
+    # Convert content to types.PromptMessage content
+    if isinstance(base_message_param.content, str):
+        contents.append(TextContent(type="text", text=base_message_param.content))
+    elif isinstance(base_message_param.content, Iterable):
+        for part in base_message_param.content:
+            if isinstance(part, TextPart):
+                contents.append(TextContent(type="text", text=part.text))
+            elif isinstance(part, ImagePart):
+                contents.append(
+                    ImageContent(
+                        type="image",
+                        data=part.image.decode("utf-8"),
+                        mimeType=part.media_type,
+                    )
+                )
+            else:
+                raise ValueError(f"Unsupported content type: {type(part)}")
+
+    else:
+        raise ValueError(
+            f"Unsupported content type: {type(base_message_param.content)}"
+        )
+
+    return [
+        PromptMessage(role=cast(Literal["user", "assistant"], role), content=content)
+        for content in contents
+    ]
+
+
+def _generate_prompt_from_function(fn: Callable) -> Prompt:
+    """
+    Generate a Prompt object from a function, extracting metadata like argument descriptions
+    from the function's docstring and type hints.
+
+    Args:
+        fn (Callable): The function to process.
+
+    Returns:
+        Prompt: A structured Prompt object with metadata.
+    """
+    # Parse the docstring for structured information
+    docstring = parse(fn.__doc__ or "")
+
+    # Extract general description from the docstring
+    description = docstring.short_description or ""
+
+    # Prepare to extract argument details
+    arguments = []
+    signature = inspect.signature(fn)
+    parameter_docs = {param.arg_name: param.description for param in docstring.params}
+
+    for parameter in signature.parameters.values():
+        # Check if the parameter has a description in the docstring
+        description = parameter_docs.get(parameter.name, "")
+
+        # Determine if the parameter is required (no default value)
+        required = parameter.default is inspect.Parameter.empty
+
+        # Add to arguments
+        arguments.append(
+            PromptArgument(
+                name=parameter.name, description=description, required=required
+            )
+        )
+
+    return Prompt(
+        name=fn.__name__,
+        description=description,
+        arguments=arguments,
+    )
+
+
+class MCPServer:
+    """MCP server implementation."""
+
+    def __init__(
+        self,
+        name: str,
+        version: str = "1.0.0",
+        tools: list[Callable | type[BaseModel] | type[AnthropicTool]] | None = None,
```

Comment:
I think this should accept only `Callable` and `BaseTool` types since we expect users to be using Mirascope (and thus could use existing tools that are written as functions or `BaseTool` definitions).